--- glib-2.28.8/gio/gzlibcompressor.c	2010-09-27 14:48:33.000000000 -0400
+++ ../../devel-i386/source/glib-2.28.8/gio/gzlibcompressor.c	2011-09-10 18:50:50.000000000 -0400
@@ -73,7 +73,7 @@
 g_zlib_compressor_set_gzheader (GZlibCompressor *compressor)
 {
   /* On win32, these functions were not exported before 1.2.4 */
-#if !defined (G_OS_WIN32) || ZLIB_VERNUM >= 0x1240
+#if ZLIB_VERNUM >= 0x1240
   const gchar *filename;
 
   if (compressor->format != G_ZLIB_COMPRESSOR_FORMAT_GZIP ||
@@ -93,7 +93,7 @@
 
   if (deflateSetHeader (&compressor->zstream, &compressor->gzheader) != Z_OK)
     g_warning ("unexpected zlib error: %s\n", compressor->zstream.msg);
-#endif /* !G_OS_WIN32 || ZLIB >= 1.2.4 */
+#endif /* ZLIB >= 1.2.4 */
 }
 
 G_DEFINE_TYPE_WITH_CODE (GZlibCompressor, g_zlib_compressor, G_TYPE_OBJECT,
--- glib-2.28.8/gio/gzlibdecompressor.c	2010-09-27 14:48:33.000000000 -0400
+++ ../../devel-i386/source/glib-2.28.8/gio/gzlibdecompressor.c	2011-09-10 18:51:53.000000000 -0400
@@ -76,7 +76,7 @@
 g_zlib_decompressor_set_gzheader (GZlibDecompressor *decompressor)
 {
   /* On win32, these functions were not exported before 1.2.4 */
-#if !defined (G_OS_WIN32) || ZLIB_VERNUM >= 0x1240
+#if ZLIB_VERNUM >= 0x1240
   if (decompressor->format != G_ZLIB_COMPRESSOR_FORMAT_GZIP)
     return;
 
